client/my-sites/patterns/controller.tsx

```diff
@@ -0,0 +1,63 @@
+import type { Context as PageJSContext } from '@automattic/calypso-router';
+import wpcom from 'calypso/lib/wp';
+import { getCurrentUserLocale, isUserLoggedIn } from 'calypso/state/current-user/selectors';
+import { getPatternSourceSiteID } from 'calypso/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/constants';
+import type { Pattern } from 'calypso/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/types';
+import LoggedInComponent from './logged-in';
+import LoggedOutComponent from './logged-out';
+
+type Next = ( error?: Error ) => void;
+
+export const fetchPatterns = ( context: PageJSContext, next: Next ) => {
+	if ( context.cachedMarkup ) {
+		return next();
+	}
+
+	const currentUserLocale = getCurrentUserLocale( context.store.getState() ) || context.lang;
+
+	const locale = currentUserLocale || 'en';
+
+	context.queryClient
+		.fetchQuery( {
+			queryKey: [ locale, 'patterns' ],
+			queryFn: () => {
+				return wpcom.req.get( `/ptk/patterns/${ locale }`, {
+					categories: 'intro',
+					per_page: '5',
+					post_type: 'wp_block',
+					site: getPatternSourceSiteID(),
+				} );
+			},
+			staleTime: Infinity,
+		} )
+		.then( ( patterns: Pattern[] ) => {
+			context.params.patterns = patterns;
+
```

Comment:
```suggestion
```
The caching strategy I mentioned previously means we don't need to store this data on `context.params`. We can use `useDotcomPatterns` in the component and rely on the `react-query` cache to pass the data from server to client.

Because the data has to be serialized before reaching the client, I wonder if the current setup doesn't also cause us to send the same data twice (the extra filesize is probably negated by gzip, but the duplication is still unnecessary).

Reply:
I'm struggling with understanding how that would work, could you help me? I've followed the approach from the `Plugins` page which also uses `react-query` in the same way. I'm also not sure if we can reuse the `useDotcomPatterns()` hook since it has some specific settings (such as `persist` set to `false`). We need to fetch the data before we render anything, does that mean I should call our new hook in the middleware in addition to the `Patterns` component itself? As it stands, this code only triggers one API call to the `/ptk/patterns` endpoint.

Reply:
I've started adding fetching in https://github.com/Automattic/wp-calypso/pull/87726, feel free to comment there if you prefer.